File: packages/core/src/lib/nodes/definitions/pytorch/input.js
/**
 * PyTorch Input Node Definition
 */
import { NodeDefinition } from '../../base';
import { BackendFramework } from '../../contracts';
export class InputNode extends NodeDefinition {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "metadata", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {
                type: 'input',
                label: 'Input',
                category: 'input',
                color: 'var(--color-gray)',
                icon: 'CircleDashed',
                description: 'Input placeholder node',
                framework: BackendFramework.PyTorch
            }
        });
        Object.defineProperty(this, "configSchema", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: [
                {
                    name: 'shape',
                    label: 'Input Shape',
                    type: 'text',
                    default: '[1, 3, 224, 224]',
                    placeholder: '[batch, channels, height, width]',
                    description: 'Input tensor dimensions as JSON array. Overridden if connected to DataLoader.'
                },
                {
                    name: 'label',
                    label: 'Custom Label',
                    type: 'text',
                    default: 'Input',
                    placeholder: 'Enter custom label...',
                    description: 'Custom label for this input node'
                },
                {
                    name: 'note',
                    label: 'Note',
                    type: 'text',
                    placeholder: 'Add notes here...',
                    description: 'Notes or comments about this input'
                }
            ]
        });
    }
    validateIncomingConnection(sourceNodeType, sourceOutputShape, targetConfig) {
        // Input nodes can only receive connections from data source nodes
        const allowedSources = ['dataloader'];
        if (!allowedSources.includes(sourceNodeType)) {
            return `Input blocks can only receive connections from data source nodes (DataLoader). Got connection from ${sourceNodeType}`;
        }
        // Connection is valid
        return undefined;
    }
    computeOutputShape(inputShape, config) {
        // If connected to a data source (DataLoader), use its shape
        if (inputShape) {
            return inputShape;
        }
        // Otherwise, use manually configured shape
        if (config.shape && typeof config.shape === 'string') {
            const dims = this.parseShapeString(config.shape);
            if (dims) {
                return { dims };
            }
        }
        // Fallback to default shape
        const defaultDims = this.parseShapeString('[1, 3, 224, 224]');
        return defaultDims ? { dims: defaultDims } : undefined;
    }
}

File: project/block_manager/services/nodes/frameworks/pytorch/nodes/input.py
"""PyTorch Input Node Definition"""

from typing import Any, Dict, List, Optional

from ....core.types import ConfigField, Framework, NodeMetadata, TensorShape
from ..base import PyTorchNodeDefinition


class InputNode(PyTorchNodeDefinition):
    """Input layer for the neural network"""

    @property
    def metadata(self) -> NodeMetadata:
        return NodeMetadata(
            type="input",
            label="Input",
            category="input",
            color="var(--color-teal)",
            icon="Download",
            description="Network input layer",
            framework=Framework.PYTORCH
        )

    @property
    def config_schema(self) -> List[ConfigField]:
        return [
            ConfigField(
                name="shape",
                label="Input Shape",
                type="text",
                default="[1, 3, 224, 224]",
                description="Input tensor shape as JSON array (e.g., [1, 3, 224, 224])"
            )
        ]

    def compute_output_shape(
        self,
        input_shape: Optional[TensorShape],
        config: Dict[str, Any]
    ) -> Optional[TensorShape]:
        # Priority: DataLoader shape > manual config > default
        if input_shape:
            return input_shape

        # Parse shape from config
        shape_str = config.get("shape", "[1, 3, 224, 224]")
        dims = self.parse_shape_string(shape_str)

        if dims:
            return TensorShape(dims=dims, description="Input tensor")

        # Fallback to default
        return TensorShape(dims=[1, 3, 224, 224], description="Input tensor")

    def validate_incoming_connection(
        self,
        source_node_type: str,
        source_output_shape: Optional[TensorShape],
        target_config: Dict[str, Any]
    ) -> Optional[str]:
        # Input nodes can only connect from DataLoader
        if source_node_type != "dataloader":
            return "Input nodes can only connect from DataLoader"
        return None
